### DISCUSSION TITLE
Any tips on optimising fps when device is in "low power mode"?

### DISCUSSION THREAD

**Discussion by xzilja** [pixijs/pixijs#7079] [Q&A] [answered]
I've got a game that runs at stable 60fps on my iPhone X (WebGL renderer), but today I noticed that when device is put into "low power mode / low battery usage" it drops fps of my game to 40 - 45. It makes sense I guess, as user is now preferring to use less resources, however this had no major effect on any other games I tested.

I'm using default pixi / Application settings at the moment and wanted to ask if there is one I could enable in order to keep pixi / webgl running at 60fps?

**Comment by themoonrat** [comment]
As you mentioned in terms of 'it makes sense'. As a user, if my phone has to go into power saving mode, I'd want the GPU to be slowed, I'd want the CPU to be slowed, and I'd want the browser to call request animation frame less frequently. 

So there is nothing 'special' to do in this scenario. You can't even detect low power mode via apis. You can optimise rendering where possible in as generic way... it's the equivalent of now playing the content on a much older phone, I guess. You could try detecting fps, and dynamically scaling down resolution if it stays low for too long a period of time?

But ultimately, don't be surprised if browsers behaviour for low power mode on certain devices / browser is 'max 30fps for rAF' in which case no amount of optimisation will help. 

I just don't think this is a scenario to expend much energy on, imo!

**Comment by xzilja** [comment]
True, I believe there is no way around this anyway, after further research it is probably related to `requestAnimationFrame` throttling in ios, it drops to 30fps.

**Comment by bigtimebuddy** [comment, marked as answer]
Maybe you're already aware, but it's important not to implement frame-based timelines in order to handle a variety of low FPS situations. If you are using PixiJS' Ticker class, the update handlers pass a scalar value you can use to multiply any time-based calculation. This is useful so that when your device enters low-power mode,  your tweens and other time-based events don't drag for users. 

In this example if we dropped to 30fps, the delta would be `2` and you'd rotate twice as fast. 

```js
// Listen for animate update
app.ticker.add((delta) => {
    // use delta to create frame-independent transform
    container.rotation -= 0.01 * delta;
});
```

**Reply by xzilja** [reply]
I see, in my case I use gsap as a ticker director as per this example https://pixijs.io/examples/#/gsap3-interaction/gsap3-tick.js
In this scenario do I need to utilise delta (I can get it from that `gsap.ticker.add` method no problem)?

```ts
app.ticker.stop();

gsap.ticker.add(() => {
  app.ticker.update();
});
```

I will read up on this topic in gsap docs as well, will come back and update this thread if I find something additional that needs to be done.

**Reply by xzilja** [reply]
In terms of gsap, nothing in particular needs to be done, it uses time based animations as opposed to ticks, related forum post
https://greensock.com/forums/topic/26597-handling-low-power-mode-when-using-gsap-as-ticker-director-with-pixi/?tab=comments#comment-129499

**Reply by bigtimebuddy** [reply]
Yes, so if you're using gasp for all your easing/tweening, then not a problem. Would be handled and interpolated internally. If you are doing any manual tweening like the example above, then you would need to get access to gsap's `deltaRatio`.
